File: Examen/Examen_1er_Bimestre/src/com/ejemplo/dao/LibroDao.java
package java.com.ejemplo.dao;

import java.com.ejemplo.model.Libro;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class LibroDao {
    private static final String FILENAME = "libros.txt";

    // Guardar un libro
    public void save(Libro libro) throws IOException {
        String data = libro.getId() + "," + libro.getTitulo() + "," + libro.getGenero() + "," + libro.getPrecio() + "," + libro.getAutor().getId();
        FileUtil.writeToFile(FILENAME, data);
    }

    // Obtener todos los libros
    public List<Libro> findAll() throws IOException {
        List<Libro> libros = new ArrayList<>();
        List<String> lines = FileUtil.readFromFile(FILENAME);
        for (String line : lines) {
            String[] parts = line.split(",");
            Libro libro = new Libro();
            libro.setId(Integer.parseInt(parts[0]));
            libro.setTitulo(parts[1]);
            libro.setGenero(parts[2]);
            libro.setPrecio(Double.parseDouble(parts[3]));
            // Aquí deberías buscar el autor por ID y asignarlo
            libros.add(libro);
        }
        return libros;
    }

    // Actualizar un libro
    public void update(Libro libro) throws IOException {
        List<Libro> libros = findAll();
        for (int i = 0; i < libros.size(); i++) {
            if (libros.get(i).getId() == libro.getId()) {
                libros.set(i, libro);
                break;
            }
        }
        saveAll(libros);
    }

    // Eliminar un libro por ID
    public void delete(int id) throws IOException {
        List<Libro> libros = findAll();
        libros.removeIf(libro -> libro.getId() == id);
        saveAll(libros);
    }

    // Obtener un libro por ID
    public Libro findById(int id) throws IOException {
        List<Libro> libros = findAll();
        for (Libro libro : libros) {
            if (libro.getId() == id) {
                return libro;
            }
        }
        return null;
    }

    // Guardar todos los libros
    private void saveAll(List<Libro> libros) throws IOException {
        StringBuilder data = new StringBuilder();
        for (Libro libro : libros) {
            data.append(libro.getId()).append(",")
                    .append(libro.getTitulo()).append(",")
                    .append(libro.getGenero()).append(",")
                    .append(libro.getPrecio()).append(",")
                    .append(libro.getAutor().getId()).append("\n");
        }
        FileUtil.writeToFile(FILENAME, data.toString(), false);
    }
}

File: Examen_02/src/dao/LibroDao.kt
package dao

import model.Autor
import model.Libro
import java.io.IOException


class LibroDao {
    private val FILENAME = "resources/libros.txt"
    private val autorDao = AutorDao()

    @Throws(IOException::class)
    private fun generateId(): Int {
        val libros = findAll()
        return (libros.maxOfOrNull { it.id } ?: 0) + 1
    }

    @Throws(IOException::class)
    fun save(libro: Libro) {
        if (libro.id == 0) {
            libro.id = generateId()
        }
        val data = "${libro.id},${libro.titulo},${libro.genero},${libro.precio},${libro.autor.id}"
        FileUtil.writeToFile(FILENAME, data)
    }

    @Throws(IOException::class)
    fun findAll(): List<Libro> {
        val libros = mutableListOf<Libro>()
        val lines = FileUtil.readFromFile(FILENAME)
        for (line in lines) {
            val parts = line.split(",")
            if (parts.size < 5) continue
            val autor = autorDao.findById(parts[4].toInt())
            if (autor != null) {
                val libro = Libro(
                    id = parts[0].toInt(),
                    titulo = parts[1],
                    genero = parts[2],
                    precio = parts[3].toDouble(),
                    autor = autor
                )
                libros.add(libro)
            }
        }
        return libros
    }

    @Throws(IOException::class)
    fun update(libro: Libro) {
        val libros = findAll().toMutableList()
        val index = libros.indexOfFirst { it.id == libro.id }
        if (index != -1) {
            libros[index] = libro
            saveAll(libros)
        }
    }

    @Throws(IOException::class)
    fun delete(id: Int) {
        val libros = findAll().toMutableList()
        libros.removeIf { it.id == id }
        saveAll(libros)
    }

    @Throws(IOException::class)
    fun findById(id: Int): Libro? {
        return findAll().find { it.id == id }
    }

    @Throws(IOException::class)
    fun findByAutorId(autorId: Int): List<Libro> {
        return findAll().filter { it.autor.id == autorId }
    }

    @Throws(IOException::class)
    private fun saveAll(libros: List<Libro>) {
        val data = libros.joinToString("\n") {
            "${it.id},${it.titulo},${it.genero},${it.precio},${it.autor.id}"
        }
        FileUtil.writeToFile(FILENAME, data, append = false)
    }
}
